model/riscv_insts_zbb.sail

```diff
@@ -0,0 +1,365 @@
+  X(rd) = EXTS(result);
+  RETIRE_SUCCESS
+}
+
+/* ****************************************************************** */
+union clause ast = ZBB_RTYPE : (regidx, regidx, regidx, brop)
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_ANDN) if haveZbb()
+  <-> 0b0100000 @ rs2 @ rs1 @ 0b111 @ rd @ 0b0110011
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_ORN) if haveZbb()
+  <-> 0b0100000 @ rs2 @ rs1 @ 0b110 @ rd @ 0b0110011
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_XNOR) if haveZbb()
+  <-> 0b0100000 @ rs2 @ rs1 @ 0b100 @ rd @ 0b0110011
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_MAX) if haveZbb()
+  <-> 0b0000101 @ rs2 @ rs1 @ 0b110 @ rd @ 0b0110011
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_MAXU) if haveZbb()
+  <-> 0b0000101 @ rs2 @ rs1 @ 0b111 @ rd @ 0b0110011
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_MIN) if haveZbb()
+  <-> 0b0000101 @ rs2 @ rs1 @ 0b100 @ rd @ 0b0110011
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_MINU) if haveZbb()
+  <-> 0b0000101 @ rs2 @ rs1 @ 0b101 @ rd @ 0b0110011
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_ROL) if haveZbb()
+  <-> 0b0110000 @ rs2 @ rs1 @ 0b001 @ rd @ 0b0110011
+
+mapping clause encdec = ZBB_RTYPE(rs2, rs1, rd, RISCV_ROR) if haveZbb()
+  <-> 0b0110000 @ rs2 @ rs1 @ 0b101 @ rd @ 0b0110011
+
+mapping zbb_rtype_mnemonic : brop <-> string = {
+  RISCV_ANDN <-> "andn",
+  RISCV_ORN  <-> "orn",
+  RISCV_XNOR <-> "xnor",
+  RISCV_MAX  <-> "max",
+  RISCV_MAXU <-> "maxu",
+  RISCV_MIN  <-> "min",
+  RISCV_MINU <-> "minu",
+  RISCV_ROL  <-> "rol",
+  RISCV_ROR  <-> "ror"
+}
+
+mapping clause assembly = ZBB_RTYPE(rs2, rs1, rd, op)
+  <-> zbb_rtype_mnemonic(op) ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+function clause execute (ZBB_RTYPE(rs2, rs1, rd, op)) = {
+  let rs1_val = X(rs1);
+  let rs2_val = X(rs2);
+  let result : xlenbits = match op {
+    RISCV_ANDN => rs1_val & ( ~ (rs2_val)),
+    RISCV_ORN  => rs1_val | ( ~ (rs2_val)),
+    RISCV_XNOR => ~ (rs1_val ^ rs2_val),
+    RISCV_MAX  => if (rs1_val <_s rs2_val)
+                    then rs2_val
+                    else rs1_val,
+    RISCV_MAXU => if (rs1_val <_u rs2_val)
+                    then rs2_val
+                    else rs1_val,
+    RISCV_MIN  => if (rs1_val <_s rs2_val)
+                    then rs1_val
+                    else rs2_val,
+    RISCV_MINU => if (rs1_val <_u rs2_val)
+                    then rs1_val
+                    else rs2_val,
+    RISCV_ROL  => if sizeof(xlen) == 32
+                    then (rs1_val << rs2_val[4..0]) | (rs1_val >> (0x20 - EXTZ(rs2_val[4..0])))
+                    else (rs1_val << rs2_val[5..0]) | (rs1_val >> (0x40 - EXTZ(rs2_val[5..0]))),
+    RISCV_ROR  => if sizeof(xlen) == 32
+                    then (rs1_val >> rs2_val[4..0]) | (rs1_val << (0x20 - EXTZ(rs2_val[4..0])))
+                    else (rs1_val >> rs2_val[5..0]) | (rs1_val << (0x40 - EXTZ(rs2_val[5..0])))
+  };
```

Comment:
I'll Update for RISCV_ROL and RISCV_ROR, rest was simpler before I think..

Reply:
I deliberately wrote it this way to match how AMOMIN[U] and AMOMAX[U] are implemented, not just because I felt like it

Reply:
Although of course s/min/max for the RISCV_MAX[U] cases, I messed that up

Reply:
Is it significant that it be done the same way as the AMOs are done? Consistency is nice, but should perhaps the AMOs change instead if the functionality is the same and it's more obvious? Is it because using the intrinsic min/max preferable (though intrinsics maxs/maxu/mins/minu might make this a no-brainer)

Reply:
I think it's much of a muchness when it comes to which is more obvious; the if/else version is the code people write in languages that lack max/min functions, and the max/min version has max/min and signed/unsigned right there in the implementation. The only ugliness is that using integers rather than bitvectors requires a to_bits on the result.

I agree that min/max_u/s operating on bitvectors would make things better; part of my original thinking that I didn't state was that, by using the min/max version, it's easier to identify the places that can be changed to use min/max_u/s if someone adds those at a later date to clean things up and hide the to_bits messiness as an implementation detail, since you can't instantly determine which ifs are acting as min/max expressions. Having said that, the instruction mnemonics have MIN and MAX in their names so they're still not hard to find, in this specific instance.